### PR TITLE
add text as parameter of onTapLink

### PR DESCRIPTION
**Pull request by bradchien**
Due to displaying title in in-app web page, it's better to know which text be clicked.

### CODE REVIEW

**Comment by mjordan56** [comment]
@bradchien I have concerns about the impact of this change on existing users of the flutter_markdown package. This change isn't backward compatible and will be a breaking change for existing users. I also believe the data you're trying to access in the createLink callback can be access by using URL parameters on the href Markdown tag for links. Using URL parameters in the href link in the Markdown and the Uri class from dart:core, you can pass text parameters with the information you want and parse the parameters using the Uri query or queryParameters methods. All of this can be accomplished without modifying the existing code and doesn't break the package for other users.

**Comment by bradchien** [comment]
@mjordan56 I totally understand your concern about backward compatible. I think using URL parameters is kind of workaround. For the document editor, they won't remember to add this extra parameter. For the developer, they have to do some trick to prevent parameter conflict and adding encode/decode URI to get link text. Adding extra parameter is more effective way to resolve this, don't you think so?

**Comment by mjordan56** [comment]
@bradchien I guess I see this PR as two distinct issues. First, the compatibility of the change, and second, the need for additional information or data to be provided to the callback method. Let's discuss these issues in order.

First, backward compatibility is not a trivial issue. Changes to a public API or library must be considerate of all the consumers. This is why care is taken to correctly assign an appropriate version number to the package and [semantic versioning](https://semver.org/) is used in Dart packages. Your PR increments the PATCH version of the flutter_markdown package so the version number goes from 0.4.3 to 0.4.4. This is not a valid incrementing of a change that will make the package incompatible with existing consumers. The Semantic Versioning specification specifies that changes that are incompatible with previous versions need to increment the MAJOR version. In this case, the version of the flutter_markdown package would need to go from 0.4.3 to 1.0.0.

While changing the version number of the flutter_markdown package to 1.0.0 could be done, I guess I would ask does this change and only this change warrant a 1.0.0 release? I guess that question would need to be put to the larger flutter_markdown user community. Do they want to see a new incompatible library with the only change being a string parameter added to Markdown link markers?

Second, the question of the need to provide additional data to the link onTap callback routine. I actually agree that having additional information beyond a basic URL could have benefits. If I had designed the MarkdownTapLinkCallback I would have probably considered something more robust like passing a dynamic object. This would be far more extensible and allow developers more flexibility in defining solutions to meet their needs. But implementing this now runs afoul of the same backward compatibility problem as your PR. When the flutter_markdown package is ready to evolve to a version 1.0.0 I would lobby for MarkdownTapLinkCallback to be changed to:
```
void MarkdownTapLinkCallback(String href, dynamic object);
```

So, instead of just saying "no, this can't be done" I did try thinking of workarounds and alternatives that could satisfy the requirement until such time as a 1.0.0 version of the flutter_markdown package is available with the preferred MarkdownTapLinkCallback. The URL parameter solution does satisfy the requirement of passing additional data to the callback routine to consume. I agree this is not as elegant a solution as my proposed dynamic object parameter but it will work. I find the argument that the document editor won't remember to add extra parameters, or that the developer will have to do tricks to prevent parameter conflicts to be pretty weak. These are steps in the development process and are easily resolved with minimal effort.

So, to answer your question, no, I do not think adding an extra parameter is a more effective way to resolve this problem given the constraint of keeping backward compatibility. When the package is opened up to move to version 1.0.0, then yes, I'd like to see a more robust MarkdownTapLinkCallback implementation but I'd prefer my definition instead of being limited to a simple string parameter.

**Comment by bradchien** [comment]
@mjordan56 First, yes. I agree with the version issue. This kind of modification should not just increase a PATCH version. Maybe change in the next major or minor version would be great.

And the second, yes, your definition is better than mine. I can't image other using scenario except string though.  

I would like to see the better implementation in the future version. Thanks for your advice and reply. :D

**Comment by mjordan56** [comment]
@bradchien I've been thinking about your PR and have some discussions with the moderator (for lack of a better term), @domesticmouse, of the flutter_markdown package. He pointed out that my assumptions regarding package incompatibility and versioning weren't completely correct. I went back and reviewed the Dart documentation on package dependencies and versioning. Dart packages that haven't declared a version 1.0.0 have more latitude to break APIs. The Dart documentation on versioning states:

> Although semantic versioning doesn’t promise any compatibility between versions prior to 1.0.0, the Dart community convention is to treat those versions semantically as well. The interpretation of each number is just shifted down one slot: going from 0.1.2 to 0.2.0 indicates a breaking change, going to 0.1.3 indicates a new feature, and going to 0.1.2+1 indicates a change that doesn’t affect the public API. For simplicity’s sake, avoid using + after the version reaches 1.0.0.

The Dart pub package manager supports this behavior so consumers of the package should be able to take new versions of the package with breaking changes in an orderly manner.

With this new understanding of allowing breaking API changes in pre-1.0.0 version packages, I'm more enthusiastic about working with you to implement an improved callback scheme for Markdown link tags. I did a deeper investigation into the issue and I have some observations and comments regarding your proposed change.

First, my proposal for changing the MarkdownTapLinkCallback method to pass a dynamic object is invalid. When I looked closer at what data is available for the method to pass back to the Markdown widget I realized the only information available is what is in the Markdown link tag from the document. So, a dynamic object doesn't make sense. What does make sense is to make all of the data in the Markdown link tag available in the callback. According to the [CommonMark](https://spec.commonmark.org/0.29/#links) and [GitHub Flavored Markdown](https://github.github.com/gfm/#links) specifications, a link contains link text, a link destination, and an optional link title. I'd like to propose that the correct change to the MarkdownTapLinkCallback be this:

```
typedef void MarkdownTapLinkCallback(String text, String href, String title);
```
The `visitElementBefore` will need to be modified to inject this additional link title information along with the link text and the link uri data.

I pulled your PR changes locally so I could look closer at your proposed changes. I have a question about the `extractTextFromElement` method. Can you explain to me why this is a recursive method? I believe I see why you want to do this, to be able to include the image description of Markdown image tags that could be included in the link text. Is this correct? If so, I think the use of a recursive method is probably excessive in this instance. I say this because the Markdown specifications state links may not contain other links at any level of nesting. When multiple valid links are nested inside each other the inner-most definition is used. The link data received in the `visitElementBefore` will be flattened so recursion isn't necessary.

The only other case I could see requiring extracting text from child elements is the case of Markdown image tags being imbedded in the link text. I can see a case for including the image description in the link text, but this text is only one level down from the parent Markdown link tag. I think there is a more straightforward method to check for Markdown image tags in the link text and composing the link text with the image description. I would propose something like this:
```
    if (tag == 'a') {
      final String text = element.children.whereType<md.Element>().isNotEmpty
          ? element.children
              .map((e) => e is md.Text
                  ? e.textContent
                  : e is md.Element && e.tag == 'img'
                      ? e.attributes['alt'] ?? ''
                      : '')
              .join('')
          : element.textContent;

      final String title = (element.attributes?.isNotEmpty ?? false)
          ? element.attributes['title']
          : '';
      _linkHandlers.add(delegate.createLink(text, element.attributes['href']));
    }
```
An appropriate comment needs to be included to explain why there is a special case for including the image description in the link text.

@bradchien what do you think? Comments?

**Comment by bradchien** [comment]
Thanks for your update.

About the recursive method `extractTextFromElement `, I indeed want to get the description for image and the nested text from the element. According to the test cases, it might have nested element mapping to one href, I am not sure if markdown only support two level nested element at that time. Now I got it. :D

Why include image descriptions in special cases? As my first comment on this thread, we need to resolved the problem when user open webpage inside the APP and not just displaying ugly url, especially those page without specific title tag in html (e.g. search link, image). Moreover, if we try to collect any analyzable events, a readable description will be more convenient.

**Comment by domesticmouse** [comment]
Thanks @bradchien for your hard work on this PR. I'm going to go ahead and land this.

File: lib/src/widget.dart [hunk 1]
```diff
@@ -17,7 +17,7 @@ import 'style_sheet.dart';
 /// Signature for callbacks used by [MarkdownWidget] when the user taps a link.
 ///
 /// Used by [MarkdownWidget.onTapLink].
-typedef void MarkdownTapLinkCallback(String href);
+typedef void MarkdownTapLinkCallback(String text, String href);
```
Comment (domesticmouse) [inline lib/src/widget.dart:20]: @bradchien @mjordan56 I just noticed this change as I was writing up the change notes. I think we should leave the href as the first parameter and add text and title after the href. 

Alternatively, as we are breaking the API anyway, I'd probably lean towards making these named parameters so it is obvious which is which. Thoughts?

**Comment by mjordan56** [comment]
Hi Brett,

Let's take your questions one at a time. First, the parameter order.
bradchien set the order of the parameters to text then href in his change.
When I first saw his PR I had the same reaction that the href parameter
should be first followed by the text parameter. But when I added the title
parameter this ordering made sense to me as there is a direct correlation
to the order of the Markdown link tag. The Markdown link tag is defined in
the GFM spec as "*A link contains link text (the visible text), a link
destination (the URI that is the link destination), and optionally a link
title.*" The format is `[*link_text*](*link_destination* "*link_title*").`
There was a symmetry ordering the parameters of the callback to match the
order of the Markdown link tag. I found this to be especially true when
writing the tests because the link title is optional and the link
destination (href) can be an empty string in a valid Markdown link. I don't
have any strong feelings about the ordering but I contend the current
ordering is as good as any.

Regarding making these named parameters, none of the other callback methods
or builder routines are using named parameters. I don't know what the
precedence is for this but I do believe in consistency. It seems like most
of the APIs I see that have required positional parameters don't use named
parameters. If you're lobbying for changing the typedef for this callback
to use named parameters then I think the other callback methods and build
routines should be named parameters as well.

Michael

**Comment by mjordan56** [comment]
@domesticmouse Let's take your questions one at a time. First, the parameter order. @bradchien set the order of the parameters to text then href in his change. When I first saw his PR I had the same reaction that the href parameter should be first followed by the text parameter. But when I added the title parameter this ordering made sense to me as there is a direct correlation to the order of the Markdown link tag. The Markdown link tag is defined in the GFM spec as "_A link contains link text (the visible text), a link destination (the URI that is the link destination), and optionally a link
title._" The format is **`[link_text](link_destination "link_title")`**. There was a symmetry ordering the parameters of the callback to match the order of the Markdown link tag. I found this to be especially true when writing the tests because the link title is optional and the link destination (href) can be an empty string in a valid Markdown link. I don't have any strong feelings about the ordering but I contend the current ordering is as good as any.

Regarding making these named parameters, none of the other callback methods or builder routines are using named parameters. I don't know what the precedence is for this but I do believe in consistency. It seems like most of the APIs I see that have required positional parameters don't use named parameters. If you're lobbying for changing the typedef for this callback to use named parameters then I think the other callback methods and build routines should be named parameters as well.

**Pull request merged**